cps/kni.c

```diff
@@ -669,6 +669,224 @@ rd_modroute(const struct nlmsghdr *req, struct route_update *update)
+{
+	struct rte_lpm6_iterator_state state6;
+	struct rte_lpm6_rule re6;
+	int index;
+
+	int ret = rte_lpm6_iterator_state_init(ltbl->lpm6, 0, 0, &state6);
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to initialize the IPv6 LPM rule iterator state in %s\n",
+			__func__);
+		return;
+	}
+
+	index = rte_lpm6_rule_iterate(&state6, &re6);
+	while (index >= 0) {
+		struct gk_fib *fib = &ltbl->fib_tbl6[re6.next_hop];
+		struct ipaddr *gw_addr;
+		struct nlmsghdr *reply;
+
+		if (fib->action != GK_FWD_GATEWAY_FRONT_NET &&
+				fib->action != GK_FWD_GATEWAY_BACK_NET &&
+				fib->action != GK_FWD_GRANTOR) {
+			index = rte_lpm6_rule_iterate(&state6, &re6);
+			continue;
+		}
+
+		reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+
+		rd_fill_getroute_reply(cps_conf, reply, fib,
+			family, req->nlmsg_seq, state6.depth, &gw_addr);
+
+		/* Add address and gateway. */
+		mnl_attr_put(reply, RTA_DST,
+			sizeof(struct in6_addr), re6.ip);
+		mnl_attr_put(reply, RTA_GATEWAY,
+			sizeof(struct in6_addr), &gw_addr->ip.v6);
+
+		if (!mnl_nlmsg_batch_next(batch)) {
+			rd_send_batch(cps_conf, batch, "IPv6",
+				req->nlmsg_seq, req->nlmsg_pid);
+			mnl_nlmsg_batch_reset(batch);
+		}
+
+		index = rte_lpm6_rule_iterate(&state6, &re6);
+	}
+}
+
+static int
+rd_getroute(const struct nlmsghdr *req)
+{
+	/*
+	 * Buffer length set according to libmnl documentation:
+	 * the buffer that you have to use to store the batch must be
+	 * double of MNL_SOCKET_BUFFER_SIZE to ensure that the last
+	 * message (message N+1) that did not fit into the batch is
+	 * written inside valid memory boundaries.
+	 */
+	char buf[2 * MNL_SOCKET_BUFFER_SIZE];
+	struct mnl_nlmsg_batch *batch;
+	struct cps_config *cps_conf = get_cps_conf();
+	struct gk_lpm *ltbl = &cps_conf->gk->lpm_tbl;
+	int family;
+
+	if (mnl_nlmsg_get_payload_len(req) < sizeof(struct rtgenmsg)) {
+		CPS_LOG(ERR, "Not enough room in CPS GETROUTE message from routing daemon in %s\n",
+			__func__);
+		goto out;
+	}
+
+	family = ((struct rtgenmsg *)mnl_nlmsg_get_payload(req))->rtgen_family;
+
+	/* We don't support AF_UNSPEC to dump both tables at once. */
+	if (family != AF_INET && family != AF_INET6) {
+		CPS_LOG(ERR, "Unsupported address family type (%d) in %s\n",
+			family, __func__);
+		goto out;
+	}
+
+	/* Leave enough room under limit for a NLMSG_DONE message. */
+	batch = mnl_nlmsg_batch_start(buf,
+		MNL_SOCKET_BUFFER_SIZE - sizeof(struct nlmsghdr));
+	if (batch == NULL) {
+		CPS_LOG(ERR, "Failed to allocate a batch for a GETROUTE reply\n");
+		goto out;
+	}
+
+	rte_spinlock_lock_tm(&ltbl->lock);
+	if (family == AF_INET)
+		rd_getroute_ipv4_locked(cps_conf, ltbl, batch, req, family);
+	else
+		rd_getroute_ipv6_locked(cps_conf, ltbl, batch, req, family);
+	rte_spinlock_unlock_tm(&ltbl->lock);
+
+	/* In the case of no entries, the only message sent is NLMSG_DONE. */
+	rd_send_batch(cps_conf, batch, family == AF_INET ? "IPv4" : "IPv6",
+		req->nlmsg_seq, req->nlmsg_pid);
```

Comment:
How does an application know when the `NLMSG_DONE` message is the last one since we send an `NLMSG_DONE` message whenever the batch buffer is full?

Reply:
Looking again, it seems that `NLMSG_DONE` should only be put at the end of the last buffer, not at the end of each buffer.